### PR TITLE
[Train] Fix multi node horovod bug

### LINKED ISSUE #20956: [Bug] [Train] Horovod trainer error in multiple nodes cluster
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

Ray Train

### What happened + What you expected to happen

I ran the example [here](https://docs.ray.io/en/master/train/examples/horovod/horovod_example.html) with default args, but it fails as following:
```
2021-12-08 10:39:50,622 INFO services.py:1338 -- View the Ray dashboard at http://127.0.0.1:8266
2021-12-08 10:39:51,863 INFO trainer.py:172 -- Trainer logs will be logged in: /home/lzhi/ray_results/train_2021-12-08_10-39-51
Traceback (most recent call last):
  File "demo_horovod.py", line 262, in <module>
    main(num_workers=args.num_workers, use_gpu=use_cuda, kwargs=kwargs)
  File "demo_horovod.py", line 147, in main
    trainer.start()
  File "/home/lzhi/miniconda3/envs/raydp/lib/python3.8/site-packages/ray/train/trainer.py", line 213, in start
    ray.get(self._backend_executor_actor.start.remote(initialization_hook))
  File "/home/lzhi/miniconda3/envs/raydp/lib/python3.8/site-packages/ray/_private/client_mode_hook.py", line 105, in wrapper
    return func(*args, **kwargs)
  File "/home/lzhi/miniconda3/envs/raydp/lib/python3.8/site-packages/ray/worker.py", line 1713, in get
    raise value.as_instanceof_cause()
ray.exceptions.RayTaskError(ValueError): ray::BackendExecutor.start() (pid=344854, ip=10.112.228.4, repr=<ray.train.backend.BackendExecutor object at 0x7f0665ae6c10>)
  File "/home/lzhi/miniconda3/envs/raydp/lib/python3.8/site-packages/ray/train/backend.py", line 177, in start
    self._backend.on_start(self.worker_group, self._backend_config)
  File "/home/lzhi/miniconda3/envs/raydp/lib/python3.8/site-packages/ray/train/horovod.py", line 143, in on_start
    nics = detect_nics(
  File "/home/lzhi/miniconda3/envs/raydp/lib/python3.8/site-packages/horovod/ray/utils.py", line 71, in detect_nics
    nics = _driver_fn(workers, all_host_names, local_host_names,
  File "/home/lzhi/miniconda3/envs/raydp/lib/python3.8/site-packages/horovod/ray/driver_service.py", line 47, in _driver_fn
    raise ValueError(f"Number of node actors ({len(node_actors)}) "
ValueError: Number of node actors (2) must match num_hosts (1).
```

### Versions / Dependencies

ray 1.9.0, horovod 0.23.0

### Reproduction script

The [hotovod_example](https://docs.ray.io/en/master/train/examples/horovod/horovod_example.html) in doc.

### Anything else

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by amogkam**
Closes https://github.com/ray-project/ray/issues/20956

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by amogkam** [comment]
Failing test before fix: https://buildkite.com/ray-project/release-tests-branch/builds/237#aa95c0da-f1fc-4224-b1da-cb8867e5a711

Passing test after fix: https://buildkite.com/ray-project/release-tests-branch/builds/240#e6f1ee95-a52d-48a4-b903-0de42a888ee3

File: release/release_tests.yaml [hunk 1]
```diff
@@ -1936,6 +1936,34 @@
 
   alert: default
 
+########################
+# Train tests
```
Comment (matthewdeng) [inline release/release_tests.yaml:1940]: Going forward, what tests should belong in `train_tests` and which tests should belong in `ml_user_tests/train`?

File: release/train_tests/horovod/train_horovod_multi_node_test.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    train_func as horovod_torch_train_func,
+)
+
+if __name__ == "__main__":
+    ray.init(address=os.environ.get("RAY_ADDRESS", "auto"))
+    start_time = time.time()
+
+    num_workers = 8
+    num_epochs = 10
+    trainer = Trainer("horovod", num_workers)
```
Comment (matthewdeng) [inline release/train_tests/horovod/train_horovod_multi_node_test.py:17]: Should we test GPU case?

**Review by matthewdeng — APPROVED** [review]
Awesome! Thanks for adding a release test as well.

File: release/release_tests.yaml (hunk 1 above)
Comment (amogkam) [reply release/release_tests.yaml:1940]: ML user tests have a specific set of properties that we want to test https://github.com/ray-project/ray/blob/ray-1.11.0/release/.buildkite/build_pipeline.py#L268.

Unfortunately this comment got removed during the refactor.

File: release/train_tests/horovod/train_horovod_multi_node_test.py (hunk 2 above)
Comment (amogkam) [reply release/train_tests/horovod/train_horovod_multi_node_test.py:17]: Doesn't seem necessary for this PR. Is there a particular behavior that you think should be tested and cannot be done with CI?

Comment (matthewdeng) [reply release/train_tests/horovod/train_horovod_multi_node_test.py:17]: Nope not in particular, just thought that between CPU and GPU we may prefer GPU! Not sure if Horovod in particular has any additional gotchas around multi-node GPU training.

Comment (amogkam) [reply release/train_tests/horovod/train_horovod_multi_node_test.py:17]: yeah unless there is a particular gotcha we want to test, I would say let's opt for the cheaper (cost-wise) test :)

Comment (amogkam) [reply release/train_tests/horovod/train_horovod_multi_node_test.py:17]: In this case, the code path we want to test will happen in any multi-node setup regardless of CPU or GPU.

**Pull request merged**